plugins/dell-dock2/fu-dell-dock2-plugin.c

```diff
@@ -0,0 +1,384 @@
+	if (vid == MST_VMM8430_USB_VID && pid == MST_VMM8430_USB_PID)
+		instance_id = g_strdup(MST_VMM8430_INST_ID);
+	if (vid == MST_VMM9430_USB_VID && pid == MST_VMM9430_USB_PID)
+		instance_id = g_strdup(MST_VMM9430_INST_ID);
+	fu_device_add_instance_id(device_mst, instance_id);
+
+	/* (TODO) remove the enforced requirement */
+	if (fu_device_has_internal_flag(device_mst, FU_DEVICE_INTERNAL_FLAG_ENFORCE_REQUIRES)) {
+		fu_device_remove_internal_flag(device_mst,
+					       FU_DEVICE_INTERNAL_FLAG_ENFORCE_REQUIRES);
+	}
+	return;
+}
+
+static void
+fu_dell_dock2_plugin_setup_relationship(FuPlugin *plugin, FuDevice *device)
+{
+	FuDevice *device_ec = fu_plugin_cache_lookup(plugin, "ec");
+	FuDevice *device_usb4 = fu_plugin_cache_lookup(plugin, "usb4");
+	FuDevice *device_mst = fu_plugin_cache_lookup(plugin, "mst");
+
+	if (device_ec && device_usb4 && !fu_device_get_parent(device_usb4)) {
+		fu_device_add_child(device_ec, device_usb4);
+		fu_plugin_cache_remove(plugin, "usb4");
+	}
+
+	if (device_ec && device_mst && !fu_device_get_parent(device_mst)) {
+		fu_dell_dock2_plugin_config_mst_dev(device_ec, device_mst);
+		fu_device_add_child(device_ec, device_mst);
+		fu_plugin_cache_remove(plugin, "mst");
+	}
+}
+
+static void
+fu_dell_dock2_plugin_device_registered(FuPlugin *plugin, FuDevice *device)
+{
+	if (fu_device_has_guid(device, DELL_DOCK2_TBT4) ||
+	    fu_device_has_guid(device, DELL_DOCK2_TBT5)) {
+		// (TODO): skip restart when passive update is ready
+		// fu_device_add_flag(device, FWUPD_DEVICE_FLAG_SKIPS_RESTART);
+		fu_plugin_cache_add(plugin, "usb4", device);
+	}
+
+	/* leverage synaptics_vmm9 plugin for mst device */
+	if (FU_IS_USB_DEVICE(device)) {
+		guint16 vid = fu_usb_device_get_vid(FU_USB_DEVICE(device));
+		guint16 pid = fu_usb_device_get_pid(FU_USB_DEVICE(device));
+
+		if ((vid == MST_VMM8430_USB_VID && pid == MST_VMM8430_USB_PID) ||
+		    (vid == MST_VMM9430_USB_VID && pid == MST_VMM9430_USB_PID)) {
+			fu_plugin_cache_add(plugin, "mst", device);
+		}
+	}
+
+	/* add ec to cache */
+	if (FU_IS_DELL_DOCK2_EC(device))
+		fu_plugin_cache_add(plugin, "ec", device);
+
+	/* usb4 device from thunderbolt plugin */
+	if (g_strcmp0(fu_device_get_plugin(device), "thunderbolt") == 0 &&
+	    fu_device_has_guid(device, DELL_DOCK2_TBT4)) {
+		g_autofree gchar *msg = NULL;
+		msg = g_strdup_printf("firmware update inhibited by [%s] plugin",
+				      fu_plugin_get_name(plugin));
+		fu_device_inhibit(device, "hidden", msg);
+		return;
+	}
+
+	/* online activation is mutually exclusive between usb4 and ec */
+	fu_dell_dock2_plugin_activate_tbt_first(plugin, NULL);
+
+	/* setup parent device */
+	fu_dell_dock2_plugin_setup_relationship(plugin, device);
+}
+
+static gboolean
+fu_dell_dock2_plugin_composite_prepare(FuPlugin *plugin, GPtrArray *devices, GError **error)
+{
+	FuDevice *ec_device = fu_plugin_cache_lookup(plugin, "ec");
+	g_autoptr(FuDeviceLocker) locker = NULL;
+
+	if (!FU_IS_DELL_DOCK2_EC(ec_device))
+		return FALSE;
+
```

Comment:
You need to set error here

Reply:
Well actually this is causing a CI failure.

You should just return TRUE if nothing to do I feel.

Reply:
ec not found yields an error which seems expected, we should stop the update.  No idea why ci complains a false return in this stage.

```
13:52:56.782 FuEngine             install phase now composite-prepare

failed to prepare composite action: failed to composite_prepare using dell_dock2: EC device not found, cannot prepare firmware updates.

13:52:56.785: Unregistered Authentication Agent for unix-process:478:42734 (system bus name :1.20, object path /org/freedesktop/PolicyKit1/AuthenticationAgent, locale en_US.UTF-8) (disconnected from bus)
No journal files were found.

-- No entries --
```

Reply:
Because composite prepare runs for all devices and plugins. It shouldn't fail unless there is a real failure.

Reply:
Thanks for the explanation. What can I do to stop this plugin? Can we place a pre-check on plugin name whether if it is `dell_dock2`?

Reply:
Composite prepare runs as part of the transaction. If you only update a device from a different plugin you might still run it?

Reply:
I've just realized that the goal can be satisfied by plugin's `->prepare` vfunc, moved to that place.